components/tikv_util/src/logger/file_log.rs

```diff
@@ -1,67 +1,198 @@
+
+    /// Execute rotation and return a renamed file path.
+    fn rotate(&mut self, file: &mut File, old_path: &Path, new_path: &Path) -> io::Result<()> {
+        file.flush()?;
+        fs::rename(old_path, new_path)?;
+
+        *file = open_log_file(old_path)?;
+        Ok(())
+    }
+
+    /// Check if the option is enabled in configuration.
+    /// Return true if the `rotator` is valid.
+    fn is_enabled(&self) -> bool;
+}
+
+/// This `FileLogger` will iterate over a series of `Rotators`,
+/// once the context trigger the `Rotator`, it will execute a rotation.
+///
+/// After rotating, the original log file would be renamed to "{original name}.{%Y-%m-%d-%H:%M:%S}".
 /// Note: log file will *not* be compressed or otherwise modified.
-pub struct RotatingFileLogger {
-    rotation_timespan: Duration,
-    next_rotation_time: DateTime<Utc>,
-    file_path: PathBuf,
+pub struct RotatingFileLogger<N>
+where
+    N: 'static + Send + Fn(&Path) -> io::Result<PathBuf>,
+{
+    path: PathBuf,
     file: File,
+    rename: N,
+    rotators: Vec<Box<dyn Rotator>>,
+}
+
+/// Builder for `RotatingFileLogger`.
+pub struct RotatingFileLoggerBuilder<N>
+where
+    N: 'static + Send + Fn(&Path) -> io::Result<PathBuf>,
+{
+    rotators: Vec<Box<dyn Rotator>>,
+    path: PathBuf,
+    rename: N,
 }
 
-impl RotatingFileLogger {
-    pub fn new(file_path: impl AsRef<Path>, rotation_timespan: Duration) -> io::Result<Self> {
-        let file_path = file_path.as_ref().to_path_buf();
-        let file = open_log_file(&file_path)?;
-        let file_attr = fs::metadata(&file_path)?;
-        let file_modified_time = file_attr.modified().unwrap().into();
-        let next_rotation_time = compute_rotation_time(&file_modified_time, rotation_timespan);
-        Ok(Self {
-            next_rotation_time,
-            file_path,
-            rotation_timespan,
-            file,
+impl<N> RotatingFileLoggerBuilder<N>
+where
+    N: 'static + Send + Fn(&Path) -> io::Result<PathBuf>,
+{
+    pub fn new(path: impl AsRef<Path>, rename: N) -> Self {
+        RotatingFileLoggerBuilder {
+            path: path.as_ref().to_path_buf(),
+            rotators: vec![],
+            rename,
+        }
+    }
+
+    pub fn add_rotator<R: 'static + Rotator>(mut self, rotator: R) -> Self {
+        if rotator.is_enabled() {
+            self.rotators.push(Box::new(rotator));
+        }
+        self
+    }
+
+    pub fn build(self) -> io::Result<RotatingFileLogger<N>> {
+        Ok(RotatingFileLogger {
+            rotators: self.rotators,
+            file: open_log_file(&self.path)?,
+            path: self.path,
+            rename: self.rename,
         })
     }
+}
+
+impl<N> Write for RotatingFileLogger<N>
+where
+    N: 'static + Send + Fn(&Path) -> io::Result<PathBuf>,
+{
+    fn write(&mut self, bytes: &[u8]) -> io::Result<usize> {
+        self.file.write(bytes)
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        for rotator in self.rotators.iter_mut() {
```

Comment:
Will other rotators be flushed?

Reply:
You should mark the function as `stateless` in your comments, otherwise there will be some bug if anyone want to add a stateful `rotator`.

Reply:
OK